pkg/controller/controller_instance.go

```diff
@@ -1238,7 +1300,6 @@ func clearServiceInstanceCurrentOperation(toUpdate *v1beta1.ServiceInstance) {
 	toUpdate.Status.CurrentOperation = ""
 	toUpdate.Status.OperationStartTime = nil
 	toUpdate.Status.AsyncOpInProgress = false
-	toUpdate.Status.OrphanMitigationInProgress = false
```

Comment:
Important change 1: Don't reset the orphan mitigation flag automatically anymore. Only explicitly reset it after we have successfully finished orphan mitigation.

Reply:
One thing to note is that there is no way to recover if the user changes the plan on the ServiceInstance while orphan mitigation is pending. The plan ref is cleared by the API server when the plan name is changed. The delete reconciliation process does not resolve plan refs (nor should it). We need some way of storing the plan that was sent in the unsuccessful provision request to make sure that we use that same plan in the orphan mitigation.

Reply:
@staebler the quick fix could be a part of #1790 
1. instead of removing a condition completely, check if the plan has changed.
2. add a check for `instance.Status.OrphanMitigationInProgress` and reject plan changes.

in the long term it would probably better to store the plan along with "in progress properties" in the status (and don't erase them until orphan mitigation has succeeded).

Reply:
@staebler the quick fix could be a part of #1790 
1. instead of removing a condition completely, check if the plan has changed.
2. add a check for `instance.Status.OrphanMitigationInProgress` and reject plan changes.

in the long term it would probably better to store the plan along with "in progress properties" in the status (and don't erase them until orphan mitigation has succeeded).

Reply:
I don't understand (1).

For (2), that is not sufficient. There is still a window where a plan change could get in and ruin orphan mitigation. 

1) Consider a ServiceInstance with an in-flight Provision request.
2) User changes plan on ServiceInstance. Orphan mitigation is not in progress, so plan change is accepted.
3) Provision request fails in a way that requires orphan mitigation.
4) Controller starts orphan mitigation.
5) Orphan mitigation reconciliation fails perpetually since there is no plan to use.

Reply:
> 2. User changes plan on ServiceInstance. Orphan mitigation is not in progress, so plan change is accepted

Currently any spec updates (inculding plan changes) are rejected if there is an operation in progress.
While I think we should stop blocking parameter updates (see #1755 and #1790), we might still reject plan updates while there is an operation or orphan mitigation in progress (as this is a rare usecase and hence not as annoying UX as blocking parameter updates).

Reply:
Sorry, I see the problem now.
For normal deletion (after instance was successfully provisioned), we already read plan from `ExternalProperties`:
[controller_instance.go#L1419-L1423](https://github.com/kubernetes-incubator/service-catalog/blob/1f60676f204ac822073acb78fdaa5d094ad7b2d1/pkg/controller/controller_instance.go#L1419-L1423)
For orphan mitigation, I think we can read it from `InProgressProperties`, just need stop overwriting them with `nil` in [controller_instance.go#L565](https://github.com/kubernetes-incubator/service-catalog/blob/1f60676f204ac822073acb78fdaa5d094ad7b2d1/pkg/controller/controller_instance.go#L565)
i.e. we can just pass existing `InProgressProperties` there:
```go
instance, err = c.recordStartOfServiceInstanceOperation(instance, v1beta1.ServiceInstanceOperationDeprovision, instance.Status.InProgressProperties)
```

@staebler Would it solve the problem?

Reply:
@staebler please review #1803 that should address this problem.